### PR TITLE
fix glib compilation with msvc-clang

### PR DESCRIPTION
**Pull request by AndreyMlashkin**
Specify library name and version:  **glib/all**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by StellaSmith** [comment]
I don't think this is correct, what if i want to cross-compile using clang targeting Windows?

**Comment by StellaSmith** [comment]
In fact, a lot of recipes wrongly assume that you want to use clang-cl when specifying `compiler=clang` and `os=Windows`.
clang can be used standalone on windows too, not just when cross-compiling.
I was pretty sure the the correct way of using clang-cl was setting `compiler=Visual Studio` and `compiler.toolchain=ClangCL`

**Comment by StellaSmith** [comment]
@SSE4 you seem to have created similar pull requests recently, thought you might want to know.

**Comment by SSE4** [comment]
> @SSE4 you seem to have created similar pull requests recently, thought you might want to know.

I don't want to repeat myself, as explanation is pretty long. consider checking out https://github.com/conan-io/conan-center-index/pull/9807 (and maybe https://github.com/conan-io/conan/issues/1839 as well).
there are multiple ways to use clang compiler on Windows (so far I tested at very least, 4 different ways, but there are more, actually).
unfortunately, conan model is not ideal and very limited, so it's hard to distinguish all of them. as a result, it's hard to make a recipe that satisfies all that use cases. from my point of view, it's better to support at least some of these use-cases in recipes than none of them at all.
maybe we can put these `clang-cl` PRs on hold, and continue discussion on how to properly model all these things in conan client itself. given 2.0 beta is going to appear in april, maybe it's finally a time to give this compiler first class support in conan. as always, suggestions are welcomed, as past discussions only led to dead end. otherwise it will be always fixiing one use-case, but breaking another use-case.

about `compiler=Visual Studio` + `compiler.toolset=ClangCL`, specifically - it has its own limitation.
1. that's very MSBuild-specific. altough it can be supported flawlessly for MSBuild-based (and CMake-based) libraries, others (meson, autotools, custom build systems, etc.) know nothing about Visual Studio toolsets and how to invoke them.
2. it seems to be impossible to model a `clang` version. toolset just says `ClangCL` and you have no idea if it was built with Clang 10 or Clang 13 or whatever. and if there are multiple installations in parallel - no way to specify an executable to use (as they all resolve to the same toolset, it will pick who knows which executable).
3. again, it assumes full `Visual Studio` installation (or at very least, `Visual Studio Build Tools`). however, some users want to use `clang-cl` standalone in docker images (with just Windows SDK for headers/libraries). some details in issue https://github.com/conan-io/conan/issues/9295.

**Comment by StellaSmith** [comment]
I would prefer going with the choice of putting these clang-cl pr/issues on hold and come up with a proper solution in Conan 2.0, due to not being a real correct way of dealing with these case as of now.

I find this very ironic, thanks to clang being so flexible we don't seem to have proper a way of handling all of it's use cases.

Another thing that i still haven't seen mentioned is that clang can be used in conjunction with mingw, and the target triplet of this configuration is `x86_64-windows-gnu`, which differs from msys' clang `x86_64-windows-msys` which in turns seems to be an alias for `x86_64-windows-msvc`

**Review by danimtb — DISMISSED** [review]

**Comment by AndreyMlashkin** [comment]
> I would prefer going with the choice of putting these clang-cl pr/issues on hold and come up with a proper solution in Conan 2.0, due to not being a real correct way of dealing with these case as of now.
> 
> I find this very ironic, thanks to clang being so flexible we don't seem to have proper a way of handling all of it's use cases.
> 
> Another thing that i still haven't seen mentioned is that clang can be used in conjunction with mingw, and the target triplet of this configuration is `x86_64-windows-gnu`, which differs from msys' clang `x86_64-windows-msys` which in turns seems to be an alias for `x86_64-windows-msvc`

is there already a concrete plan for conan 2.0?

**Comment by uilianries** [comment]
@AndreyMlashkin there is a conflict on this PR. Please, fix.

File: recipes/glib/all/conanfile.py [hunk 1]
```diff
@@ -39,6 +39,19 @@ class GLibConan(ConanFile):
...
+    @property
+    def _is_msvc(self):
+        return self.settings.compiler == "Visual Studio" or self.settings.compiler == "msvc"
+
+    def validate(self):
+        if hasattr(self, 'settings_build') and tools.cross_building(self, skip_x64_x86=True):
+            raise ConanInvalidConfiguration("Cross-building not implemented")
+        if tools.Version(self.version) >= "2.69.0" and not self.options.with_pcre:
+            raise ConanInvalidConfiguration("option glib:with_pcre must be True for glib >= 2.69.0")
+
```
Comment (ericLemanissier) [inline recipes/glib/all/conanfile.py:54]:
```suggestion
```

File: recipes/glib/all/conanfile.py [hunk 2]
```diff
@@ -176,7 +189,7 @@ def build(self):
         self._patch_sources()
         with tools.environment_append(
             VisualStudioBuildEnvironment(self).vars
-        ) if is_msvc(self) else tools.no_op():
+        ) if self._is_msvc or self._is_clang_cl else tools.no_op():
```
Comment (ericLemanissier) [inline recipes/glib/all/conanfile.py:192]:
```suggestion
        ) if is_msvc(self) or self._is_clang_cl else tools.no_op():
```

File: recipes/glib/all/conanfile.py [hunk 3]
```diff
@@ -192,7 +205,7 @@ def package(self):
         self.copy(pattern="COPYING", dst="licenses", src=self._source_subfolder)
         with tools.environment_append(
             VisualStudioBuildEnvironment(self).vars
-        ) if is_msvc(self) else tools.no_op():
+        ) if self._is_msvc or self._is_clang_cl else tools.no_op():
```
Comment (ericLemanissier) [inline recipes/glib/all/conanfile.py:208]:
```suggestion
        ) if is_msvc(self) or self._is_clang_cl else tools.no_op():
```

**Review by ericLemanissier — CHANGES_REQUESTED** [review]
this fixes some conflicts.
however, I don't think this should be merged before https://github.com/conan-io/conan/issues/10955 is fixed, because "clang on windows" is not sufficient information to decide that we need to use VisualStudioBuildEnvironment

File: recipes/glib/all/conanfile.py [hunk 4]
```diff
@@ -39,6 +39,8 @@ class GLibConan(ConanFile):
     generators = "pkg_config"
 
     @property
+    def _is_clang_cl(self):
+        return self.settings.os == 'Windows' and self.settings.compiler == 'clang'
```
Comment (ericLemanissier) [inline recipes/glib/all/conanfile.py:43]:
```suggestion
        return self.settings.os == 'Windows' and self.settings.compiler == 'clang'

    @property
```

**Review by prince-chrismc — APPROVED** [review]

**Comment by ericLemanissier** [comment]
@AndreyMlashkin please change the PR status to draft

**Comment by prince-chrismc** [comment]
https://github.com/conan-io/conan-center-index/pull/10029#issuecomment-1083548081

> from my point of view, it's better to support at least some of these use-cases in recipes than none of them at all.

I am in favor of a small change as long as it does not break the main workflows.

We can always circle back to make it better

**Comment by ericLemanissier** [comment]
There is no improvement here. It breaks one configuration to fix another

**Pull request closed without merging**